text/0000-list-rendering-reform.md

```diff
@@ -0,0 +1,292 @@
+
+## Alternatives
+
+### Enhancing `for:each`
+
+One reasonable alternative would be to enhance the existing `for:each` rather than creating a new directive. However, this has several drawbacks:
+
+1. There could be conflicts between the legacy `key` on within-iterator elements versus the `lwc:key` on the iterator element. In cases of nested iterators, it could be very tricky to determine which keys should apply to which iterators.
+2. It does not move us closer to a world where all LWC directives use the `lwc:` prefix. Directives like `for:each`, `for:item`, `for:index`, and `key` would continue to be the way to write iterators, making it difficult to support a [directive shorthand](https://github.com/salesforce/lwc/issues/3303) in the future.
+
+### Enhancing `iterator:*`
+
+Similarly, enhancing `iterator:*` is another approach. However, this directive is seldom-used, and is rigid in how it allows for referencing the value, index, and first and last items in a list (the property names must always be `value`, `index`, `first`, and `last`).
+
+### Making `lwc:key` optional
+
+The key is a performance optimization, and not all developers need it. Some developers even use a getter with `Math.random()` to provide a random key, to work around the current LWC requirement that lists must have a `key`.
+
+It is entirely possible to make `lwc:key` optional. However, this comes with a few downsides:
+
+1. Developers may be encouraged to use a less-performant rendering pattern, since it takes less effort to omit the key.
+2. It deviates from the existing behavior of `for:each`, and is one more thing to teach.
+3. Developers may now assume that keys are actually a _bad_ idea, since the "improved" `lwc:each` makes them optional.
+
+So, for this proposal, keys are still required. However, it is definitely something that can be revisited in the future.
+
+## Adoption strategy
+
+Similar to `lwc:if`/`lwc:elseif`/`lwc:else`, the messaging will be that `lwc:each` is the "new" way to use lists in LWC templates, and component authors should migrate from `for:each` and `iterator:*`.
+
+Unlike `lwc:if` and friends, there are no feature gaps between `lwc:each` and the existing `for:each` and `iterator:*` directives. (`lwc:if` is missing an equivalent to `if:false`, at least in the absence of complex expressions.) Therefore, it should be relatively straightforward to migrate existing components – it could even be done with [lwc-codemod](https://github.com/salesforce/lwc-codemod).
+
+In the future, we could also use [component-level API versioning](https://lwc.dev/guide/versioning#component-level-api-versioning) to disallow the legacy `for:each` and `iterator:*` directives.
+
+# How we teach this
+
+The new `lwc:each` directive is essentially a superset of `for:each`, so it can leverage most of the teaching material for that directive.
+
+In addition, we can downplay or remove documentaiton for `for:each` and `iterator:*`, simplifying the LWC education narrative.

```

Comment:
I think it might be a _bit_ early to do this. Similar to `if:true` and `if:false`, we need to make sure first that we have a well-lit migration path. But we can figure this out later.